airflow/www/views.py

```diff
@@ -2523,6 +2523,48 @@ def action_delete(self, ids):
             super(TaskInstanceModelView, self).action_delete(ids)
 
     @provide_session
+    @action('clear',
+            lazy_gettext('Clear'),
+            lazy_gettext(
+                'Are you sure you want to clear the state of task instance(s) associated with selected task(s)?\n'
+                'NOTE: DAG state(s) will be set to "RUNNING".'))
+    def action_clear(self, ids, session=None):
+        try:
+            TI = models.TaskInstance
+
+            dag_ti_mapping = {}
+            drs = []
+
+            for id in ids:
+                task_id, dag_id, execution_date = id.split(',')
+
+                ti = session.query(TI).filter(TI.task_id == task_id,
+                                              TI.dag_id == dag_id,
+                                              TI.execution_date == execution_date).one()
+
+                dag = dagbag.get_dag(dag_id)
+                drs.extend(session.query(models.DagModel).filter_by(dag_id=dag_id).all())
+                ti_list = dag_ti_mapping.setdefault(dag, [])
```

Comment:
We can simplify this a bit, make a dict from dag id to list of task instances, which is a default dict to the empty list. Then you can just do a dag_to_cleared_tis[dag_id].append(ti)